sdk.go

```diff
@@ -0,0 +1,162 @@
+// return tuple, the only place for errors now is in the Response type.
+// I think this makes good sense. Lets not create a failure condition
+// from something that could be in user space.
+type Handler interface {
+	Handle(ctx context.Context, r Request) Response
+}
+
+type HandlerFn func(ctx context.Context, r Request) Response
+
+func (h HandlerFn) Handle(ctx context.Context, r Request) Response {
+	return h(ctx, r)
+}
+
+// HandleFnOf provides a means to translate the incoming requests to the destination body type.
+// This normalizes the sad path and provides the caller with a zero fuss request to work with. Reducing
+// json boilerplate for what is essentially the same operation on different types.
+//
+// TODO(berg): name could be HandlerOf perhaps?
+func HandleFnOf[T any](fn func(context.Context, RequestOf[T]) Response) Handler {
+	return HandlerFn(func(ctx context.Context, r Request) Response {
+		var v T
+		if err := json.Unmarshal(r.Body, &v); err != nil {
+			return Response{Errors: []APIError{{Code: http.StatusBadRequest, Message: "failed to unmarshal payload: " + err.Error()}}}
+		}
+
+		return fn(ctx, RequestOf[T]{
+			Body:        v,
+			Context:     r.Context,
+			Params:      r.Params,
+			URL:         r.URL,
+			Method:      r.Method,
+			AccessToken: r.AccessToken,
+		})
+	})
+}
+
+type (
+	// Request a word for Request. We can treat it as an internal impl details, and provide the
+	// user an http.Request, built from this internal impl. They can then build up
+	// the usual http.ServeHTTP, the same as they are likely accustomed too. We too
+	// can take advantage of this and add middleware,etc to the runner as needed. When testing
+	// locally, they just use curl, and they provide it the way they are use too. For our lambda
+	// impl, we convert it to the expected request body the http server expects, then its treated
+	// the same as our standard impl.
+	Request RequestOf[json.RawMessage]
+
+	// RequestOf provides a generic body we can target our unmarshaling into. We don't have to have this
+	// as some users may be happy with the OG Request. However, we can express the former with the latter
+	// here.
+	RequestOf[T any] struct {
+		Body T
+		// TODO(berg): can we axe Context? have workflow put details in the body/headers/params instead?
+		Context json.RawMessage
+		Params  struct {
+			Header http.Header
+			Query  url.Values
+		}
+		// TODO(berg): explore changing this field to Path, as URL is misleading. It's never
+		// 			   an fqdn, only the path of the url.
+		URL         string
+		Method      string
+		AccessToken string
+	}
+
+	Response struct {
+		Body   json.Marshaler
+		Code   int
+		Errors []APIError
+		Header http.Header
+	}
+
+	APIError struct {
+		Code    int    `json:"code"`
+		Message string `json:"message"`
+	}
+)
+
+// Run is the meat and potatoes. This is the entrypoint for everything.
+func Run[T Cfg](ctx context.Context, newHandlerFn func(cfg T) Handler) {
```

Comment:
One thing I'm still contemplating here..... is if we pass context to the `newHandlerFn`. making it something like:

```
func Run[T Cfg](ctx context.Context, newHandlerFn func(ctx context.Context, cfg T) Handler) {
```

Reply:
I've convinced myself, going to add the context to the newHandler Fn. Simple change, but makes us robust to nearly any initialization an author has to do.